File: electron/fileOperations.ts
import { app } from 'electron'
import fs from 'fs/promises'
import path from 'path'
import type { AppSettings, HistoryEntry, FolderValidation, KeynoteMetadata, TokenDetection } from '../src/types/index'

const DEFAULT_SETTINGS: AppSettings = {
  vercelToken: '',
  vercelTeamId: 'team_E1wAzl9zyAPrlGzyjmcXNuxd',
  theme: 'system',
  autoCopyUrl: true,
  projectNamePrefix: '',
  lastFolderPath: '',
}

function getDataDir(): string {
  return path.join(app.getPath('userData'))
}

function getSettingsPath(): string {
  return path.join(getDataDir(), 'settings.json')
}

function getHistoryPath(): string {
  return path.join(getDataDir(), 'history.json')
}

export async function loadSettings(): Promise<AppSettings> {
  try {
    const data = await fs.readFile(getSettingsPath(), 'utf-8')
    return { ...DEFAULT_SETTINGS, ...JSON.parse(data) }
  } catch {
    return { ...DEFAULT_SETTINGS }
  }
}

export async function saveSettings(settings: Partial<AppSettings>): Promise<void> {
  const current = await loadSettings()
  const merged = { ...current, ...settings }
  await fs.mkdir(getDataDir(), { recursive: true })
  await fs.writeFile(getSettingsPath(), JSON.stringify(merged, null, 2), 'utf-8')
}

export async function loadHistory(): Promise<HistoryEntry[]> {
  try {
    const data = await fs.readFile(getHistoryPath(), 'utf-8')
    return JSON.parse(data)
  } catch {
    return []
  }
}

export async function addHistoryEntry(entry: HistoryEntry): Promise<void> {
  const history = await loadHistory()
  history.unshift(entry)
  await fs.mkdir(getDataDir(), { recursive: true })
  await fs.writeFile(getHistoryPath(), JSON.stringify(history, null, 2), 'utf-8')
}

export async function validateKeynoteFolder(folderPath: string): Promise<FolderValidation> {
  try {
    const headerPath = path.join(folderPath, 'assets', 'header.json')
    const mainJsPath = path.join(folderPath, 'assets', 'player', 'main.js')

    // Check header.json exists
    try {
      await fs.access(headerPath)
    } catch {
      return { valid: false, folderPath, error: 'Missing assets/header.json — not a Keynote HTML export' }
    }

    // Check main.js exists
    try {
      await fs.access(mainJsPath)
    } catch {
      return { valid: false, folderPath, error: 'Missing assets/player/main.js — not a Keynote HTML export' }
    }

    // Parse header.json for metadata
    const headerData = await fs.readFile(headerPath, 'utf-8')
    const header = JSON.parse(headerData)

    const metadata: KeynoteMetadata = {
      title: header.title || header.name || path.basename(folderPath),
      slideCount: header.slideCount || header.slides?.length || 0,
      width: header.width || 1920,
      height: header.height || 1080,
      raw: header,
    }

    return { valid: true, folderPath, metadata }
  } catch (error) {
    return { valid: false, folderPath, error: `Validation failed: ${String(error)}` }
  }
}

export async function detectVercelToken(): Promise<TokenDetection> {
  const locations = [
    path.join(process.env.HOME || '', '.local', 'share', 'com.vercel.cli', 'auth.json'),
    path.join(process.env.HOME || '', 'Library', 'Application Support', 'com.vercel.cli', 'auth.json'),
  ]

  for (const loc of locations) {
    try {
      const data = await fs.readFile(loc, 'utf-8')
      const parsed = JSON.parse(data)
      if (parsed.token) {
        return { found: true, token: parsed.token, source: loc }
      }
    } catch {
      // Try next location
    }
  }

  return { found: false }
}

File: swift-app/Sources/Services/FileOperations.swift
import Foundation

/// Settings persistence and folder validation — mirrors electron/fileOperations.ts
enum FileOperations {

    /// Application Support directory for settings
    static var dataDirectory: URL {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("keynote-deployer")
    }

    private static var settingsPath: URL { dataDirectory.appendingPathComponent("settings.json") }

    // MARK: - Settings

    static func loadSettings() throws -> AppSettings {
        guard FileManager.default.fileExists(atPath: settingsPath.path) else {
            return .default
        }
        let data = try Data(contentsOf: settingsPath)
        let saved = try JSONDecoder().decode(AppSettings.self, from: data)
        return saved
    }

    static func saveSettings(_ settings: AppSettings) throws {
        try FileManager.default.createDirectory(at: dataDirectory, withIntermediateDirectories: true)
        let data = try JSONEncoder().encode(settings)
        try data.write(to: settingsPath)
    }

    // MARK: - Folder Validation

    struct FolderValidation: Sendable {
        let valid: Bool
        let folderPath: String
        let metadata: KeynoteMetadata?
        let error: String?
    }

    static func validateKeynoteFolder(_ folderPath: String) throws -> FolderValidation {
        let fm = FileManager.default
        let headerPath = (folderPath as NSString).appendingPathComponent("assets/header.json")
        let mainJsPath = (folderPath as NSString).appendingPathComponent("assets/player/main.js")

        guard fm.fileExists(atPath: headerPath) else {
            return FolderValidation(valid: false, folderPath: folderPath, metadata: nil, error: "Missing assets/header.json — not a Keynote HTML export")
        }

        guard fm.fileExists(atPath: mainJsPath) else {
            return FolderValidation(valid: false, folderPath: folderPath, metadata: nil, error: "Missing assets/player/main.js — not a Keynote HTML export")
        }

        let headerData = try Data(contentsOf: URL(fileURLWithPath: headerPath))
        let metadata = try JSONDecoder().decode(KeynoteMetadata.self, from: headerData)

        return FolderValidation(valid: true, folderPath: folderPath, metadata: metadata, error: nil)
    }

    // MARK: - Token Detection

    struct TokenDetection: Sendable {
        let found: Bool
        let token: String?
        let source: String?
    }

    static func detectVercelToken() -> TokenDetection {
        let home = NSHomeDirectory()
        let locations = [
            "\(home)/.local/share/com.vercel.cli/auth.json",
            "\(home)/Library/Application Support/com.vercel.cli/auth.json",
        ]

        for loc in locations {
            guard let data = try? Data(contentsOf: URL(fileURLWithPath: loc)),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let token = json["token"] as? String else { continue }
            return TokenDetection(found: true, token: token, source: loc)
        }

        return TokenDetection(found: false, token: nil, source: nil)
    }
}
